src/main/java/org/folio/service/mapping/processor/RuleProcessor.java

```diff
@@ -0,0 +1,103 @@
+package org.folio.service.mapping.processor;
+
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import org.folio.service.mapping.processor.rule.Rule;
+import org.folio.service.mapping.processor.rule.Translation;
+import org.folio.service.mapping.processor.translations.Settings;
+import org.folio.service.mapping.processor.translations.TranslationFunction;
+import org.folio.service.mapping.processor.translations.TranslationsHolder;
+import org.folio.service.mapping.reader.EntityReader;
+import org.folio.service.mapping.reader.values.CompositeValue;
+import org.folio.service.mapping.reader.values.ListValue;
+import org.folio.service.mapping.reader.values.RuleValue;
+import org.folio.service.mapping.reader.values.SimpleValue;
+import org.folio.service.mapping.reader.values.StringValue;
+import org.folio.service.mapping.writer.RecordWriter;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import static org.folio.service.mapping.reader.values.SimpleValue.SubType.LIST_OF_STRING;
+import static org.folio.service.mapping.reader.values.SimpleValue.SubType.STRING;
+
+/**
+ * RuleProcessor is a central part of mapping.
+ *
+ * High-level algorithm:
+ * # read data by the given rule
+ * # translate data
+ * # write data
+ *
+ * @see EntityReader
+ * @see TranslationFunction
+ * @see RecordWriter
+ */
+public final class RuleProcessor {
+  private List<Rule> rules = new ArrayList<>();
+
+  public RuleProcessor(JsonArray rules) {
+    Iterator ruleIterator = rules.iterator();
+    while (ruleIterator.hasNext()) {
+      Rule rule = new Rule(JsonObject.mapFrom(ruleIterator.next()));
+      this.rules.add(rule);
+    }
+  }
+
+  public String process(EntityReader reader, RecordWriter writer, Settings settings) {
+    this.rules.forEach(rule -> {
+      RuleValue ruleValue = reader.read(rule);
+      switch (ruleValue.getType()) {
```

Comment:
The idea for refactoring - implement design pattern Command instead of using a switch operator. For example, you can create two classes SimpleRuleProcessor and CompositeRuleProcessor and move there translate(...) methods.  Static final Map can be used o initialize and use such methods

Reply:
Good idea but let's keep it simple at this moment and leave as it is